### PR TITLE
Backend shade layer

### LINKED ISSUE #530: Security/Performancen testing strategy
We miss two short strategies for these testing fields.
After #519 is merged, u will find the updated strategy in docs/tests/README.md

### LINKED ISSUE #638: shade layer endpoint
### Tasks

 - [ ] create table for shade layer
 - [ ] create end point that allows to add, edit, move and remove and delete areas/polygons where and how much shade is (matching to enum existing already for plants)
 - [ ] use shade layer in heatmap

### Use case

shade_layer

### Related Pull request

_No response_

### LINKED ISSUE #650: Heatmap Test Protocol
I think we all have small misunderstandings about the manual test protocol, I hope I can clear this up on tuesday.
@markus2330 @kitzbergerg 

Manual tests should also be executable without the need of such a big json string. If every test requires that, we will have a hundred page long test protocol. Additionally the terminal should not be necessary for them to be performed. This increases the duration to perform the test.

[This](https://github.com/ElektraInitiative/PermaplanT/issues/640#:~:text=The%20general%20idea,the%20process%20clearer.)

### PR DESCRIPTION
**Pull request by kitzbergerg**
Resolves #638, resolves #650, resolves #530.

## Basics

- [x] I added a line to [/doc/CHANGELOG.md](/doc/CHANGELOG.md)
- [x] The PR is rebased with current master.
- [x] Details of what you changed are in commit messages.
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildserver is happy.

## Checklist

- [ ] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](https://github.com/ElektraInitiative/PermaplanT/blob/master/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

### CODE REVIEW

**Review by badnames — COMMENTED** [review]
The API looks good thus far.

I can add a more detailed review once this PR is no longer a draft.

File: backend/migrations/2023-07-22-110000_shadings/up.sql [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    -- Create a point from x_pos and y_pos
+    point := ST_SetSRID(ST_MakePoint(x_pos, y_pos), 4326);
+    -- Select the shading with the darkest shade that intersects the point
+    SELECT shade INTO shading_shade
+    FROM shadings
+    WHERE layer_id = p_layer_id AND ST_Intersects(geometry, point)
+    ORDER BY shade DESC
+    LIMIT 1;
+
+    -- If there's no shading, return 0
```
Comment (markus2330) [inline backend/migrations/2023-07-22-110000_shadings/up.sql:79]: Do I understand correctly that the places where nothing was drawn on the shade layer, we will discourage any plant from being placed there?

**Review by markus2330 — CHANGES_REQUESTED** [review]
Can you please split the changes of

- [backend/migrations/2023-07-03-165000_heatmap/up.sql](https://github.com/ElektraInitiative/PermaplanT/pull/684/files#diff-90e920f3ee01f0da92f3e2f47880632407ab1bba770c44355e9bd0c864877896), I want to merge this now as we need this for the stable release (after that we don't want any incompatible changes in migrations anymore, i.e. no modification of existing migration files)
- improvements of patch file (there is also #676, maybe simply push it there)
- and the actual shade layer

@badnames can you please test the shadings API now, it is not draft anymore?

File: backend/migrations/2023-07-22-110000_shadings/up.sql (hunk 1 above)
Comment (kitzbergerg) [reply backend/migrations/2023-07-22-110000_shadings/up.sql]: This mean shadings is not taken into account at all. The score returned from this function is between -0.5 and 0.5

**Comment by kitzbergerg** [comment]
> Can you please split the changes of
> 
> * [backend/migrations/2023-07-03-165000_heatmap/up.sql](https://github.com/ElektraInitiative/PermaplanT/pull/684/files#diff-90e920f3ee01f0da92f3e2f47880632407ab1bba770c44355e9bd0c864877896), I want to merge this now as we need this for the stable release (after that we don't want any incompatible changes in migrations anymore, i.e. no modification of existing migration files)
> * improvements of patch file (there is also [659 improve documentation about patch file #676](https://github.com/ElektraInitiative/PermaplanT/pull/676), maybe simply push it there)
> * and the actual shade layer

Do you mean to split this PR into 3 separate PRs? If so that won't really be possible as the changes are all interlinked (e.g. I would need to change the Rust code as well to fit with the updated heatmap SQL, which would mean somehow extracting all relevant changes from this PR).
Since this PR is finished from my side it would be faster to merge this as is.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Looks like you've put a lot of effort into here!

Unfortunately, it is difficult to see without test protocol/images/explanations of the overall algorithm if it actually does what it should do. Some examples would be also very helpful, e.g. what is the result of the score and the color if the shading is N steps away from what the plant wants. Also test cases for that would be helpful.

> Do you mean to split this PR into 3 separate PRs? If so that won't really be possible as the changes are all interlinked (e.g. I would need to change the Rust code as well to fit with the updated heatmap SQL, which would mean somehow extracting all relevant changes from this PR).
> Since this PR is finished from my side it would be faster to merge this as is.

I am afraid it won't be so fast, as @badnames should also review etc.

I would like to do the stable release latest on Tuesday, so that we can start the user study. The shade layer is not part of the user study anyway, so there is no benefit of adding it.

But the [backend/migrations/2023-07-03-165000_heatmap/down.sql](https://github.com/ElektraInitiative/PermaplanT/pull/684/files#diff-40bd8b4c1b9dd94cac2b085de8d70ea2a4e7b390da815342aab9145c28ab5d54) and [backend/migrations/2023-07-03-165000_heatmap/up.sql](https://github.com/ElektraInitiative/PermaplanT/pull/684/files#diff-90e920f3ee01f0da92f3e2f47880632407ab1bba770c44355e9bd0c864877896) must be added before the stable release, otherwise you would need to rewrite them as a separate migration which would be ugly and extra effort for you.

So just to be safe, it would be great to have an extra PR which just includes the [backend/migrations/2023-07-03-165000_heatmap/down.sql](https://github.com/ElektraInitiative/PermaplanT/pull/684/files#diff-40bd8b4c1b9dd94cac2b085de8d70ea2a4e7b390da815342aab9145c28ab5d54) and [backend/migrations/2023-07-03-165000_heatmap/up.sql](https://github.com/ElektraInitiative/PermaplanT/pull/684/files#diff-90e920f3ee01f0da92f3e2f47880632407ab1bba770c44355e9bd0c864877896), and the rest of your changes on top of it.

If everything is ready before we release, great, then of course I will merge everything.

File: backend/migrations/2023-07-22-110000_shadings/up.sql [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    -- Create a point from x_pos and y_pos
+    point := ST_SetSRID(ST_MakePoint(x_pos, y_pos), 4326);
+    -- Select the shading with the darkest shade that intersects the point
+    SELECT shade INTO shading_shade
+    FROM shadings
+    WHERE layer_id = p_layer_id AND ST_Intersects(geometry, point)
+    ORDER BY shade DESC
+    LIMIT 1;
+
+    -- If there's no shading, return 0 (meaning shadings do not affect the score)
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply backend/migrations/2023-07-22-110000_shadings/up.sql:79]: This is not feasible given the little time left. I would have to rewrite the whole heatmap calculation including the backend.
As far as I understand this would require two values to be returned to the backend, one score values for red-green (same as it currently is) and an additional value for the alpha channel (there is nothing like this in place).
I can create a left over issue for that.

File: frontend/src/features/maps/routes/MapCreateForm.tsx [hunk 3]
```diff
@@ -130,9 +130,9 @@ export default function MapCreateForm() {
         rings: [
           [
             { x: 0.0, y: 0.0 },
-            { x: 0.0, y: 10_000.0 },
-            { x: 10_000.0, y: 10_000.0 },
-            { x: 10_000.0, y: 0.0 },
+            { x: 0.0, y: 1_000.0 },
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply frontend/src/features/maps/routes/MapCreateForm.tsx:133]: This is now 10x10m if I'm not mistaken. The 100x100m square as a default would take to long to calculate with granularity 10cm (at least it would take to long if everybody had that set as a default).
Since a default of 10x10 is more performant (and probably more realistic for most people anyways) I reduced it.

**Comment by kitzbergerg** [comment]
> Unfortunately, it is difficult to see without test protocol/images/explanations of the overall algorithm if it actually does what it should do. Some examples would be also very helpful, e.g. what is the result of the score and the color if the shading is N steps away from what the plant wants. Also test cases for that would be helpful.

I now added a manual test to show it works and so you can replicate it.

> So just to be safe, it would be great to have an extra PR which just includes the [backend/migrations/2023-07-03-165000_heatmap/down.sql](https://github.com/ElektraInitiative/PermaplanT/pull/684/files#diff-40bd8b4c1b9dd94cac2b085de8d70ea2a4e7b390da815342aab9145c28ab5d54) and [backend/migrations/2023-07-03-165000_heatmap/up.sql](https://github.com/ElektraInitiative/PermaplanT/pull/684/files#diff-90e920f3ee01f0da92f3e2f47880632407ab1bba770c44355e9bd0c864877896), and the rest of your changes on top of it.

I would basically need to rewrite this whole PR for that which is why I said it is faster to merge this. I have time this evening, tomorrow evening and on Tuesday to fix problems with this so I would ask @badnames to add his review tomorrow if that is possible.

File: backend/migrations/2023-07-22-110000_shadings/up.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply backend/migrations/2023-07-22-110000_shadings/up.sql:85]: *I'll give it a quick try. If its not to difficult I'll include it.

[the remark this answers is not in the document]
Comment (markus2330) [reply backend/migrations/2023-07-22-110000_shadings/up.sql:85]: Thx! Actually, I don't think two values would be needed, as the red is static? Having two values probably would be more flexible for future extensions, though.

File: frontend/src/features/maps/routes/MapCreateForm.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply frontend/src/features/maps/routes/MapCreateForm.tsx:133]: That is too small, even for our garden. Our garden has 200m length.

If there are unsolvable performance issues, please change the granularity of the blocks to 1m.

File: backend/migrations/2023-07-22-110000_shadings/up.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply backend/migrations/2023-07-22-110000_shadings/up.sql:86]: Done.

File: backend/migrations/2023-07-22-110000_shadings/up.sql [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+CREATE TRIGGER check_shade_layer_type_before_insert_or_update
+BEFORE INSERT OR UPDATE ON shadings
+FOR EACH ROW EXECUTE FUNCTION check_shade_layer_type();
+
+-- Calculate a score for a certain position.
+-- p_layer_ids[1] ... plant layer
+-- p_layer_ids[2] ... shade layer
+CREATE OR REPLACE FUNCTION calculate_score(
+    p_map_id integer,
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply backend/migrations/2023-07-22-110000_shadings/up.sql:35]: This seems to be automatically lowercased by the `sql format` git commit hook. So I can't even change it any more.
@4ydan Do you know what's going on here?

[the remark this answers is not in the document]
Comment (4ydan) [reply backend/migrations/2023-07-22-110000_shadings/up.sql:35]: Sadly I have not introduced that hook and dont know whats going on.

[the remark this answers is not in the document]
Comment (markus2330) [reply backend/migrations/2023-07-22-110000_shadings/up.sql:35]: @temmey do you know why it is inconsistent and how to change it?

[the remark this answers is not in the document]
Comment (temmey) [reply backend/migrations/2023-07-22-110000_shadings/up.sql:35]: Why is it inconsistent? I don't know; it works for me. But I created a pr #695 so, in the future, this won't happen anymore.

[the remark this answers is not in the document]
Comment (kitzbergerg) [reply backend/migrations/2023-07-22-110000_shadings/up.sql]: I set the pre-commit hook for SQL types to uppercase as well. Now it works.

File: backend/migrations/2023-07-22-110000_shadings/up.sql [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+END;
+$$ LANGUAGE plpgsql;
+
+-- Calculate preference: Between -0.5 and 0.5 depending on shadings.
+-- Calculate relevance: 0.5 if there is shading; otherwise 0.0.
+--
+-- If the plant would die at the position set preference=-100 and relevance=100.
+CREATE FUNCTION calculate_score_from_shadings(
+    p_layer_id INTEGER, p_plant_id INTEGER, x_pos INTEGER, y_pos INTEGER
+)
```
Comment (markus2330) [inline backend/migrations/2023-07-22-110000_shadings/up.sql:63]: I miss the date from which the shadings are used.

File: backend/migrations/2023-07-03-165000_heatmap/up.sql [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+-- (0,0) no matter the boundaries of the map.
+-- To get valid coordinates the user would therefore need to move and scale the
+-- calculated heatmap by taking into account the boundaries of the map.
+--
+-- p_map_id                ... map id
+-- p_layer_ids             ... ids of the layers
+-- p_plant_id              ... id of the plant for which to consider relations
+-- granularity             ... resolution of the map (must be greater than 0)
+-- x_min,y_min,x_max,y_max ... boundaries of the map
+CREATE OR REPLACE FUNCTION calculate_heatmap(
```
Comment (markus2330) [inline backend/migrations/2023-07-03-165000_heatmap/up.sql:48]: I miss the date of for which data the heatmap gets generated

File: benchmarks/backend/config/setup.sh [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
```
Comment (markus2330) [inline benchmarks/backend/config/setup.sh:2]: Setup of backend?

File: benchmarks/backend/heatmap/insert_data-small_map.sh [hunk 8]
```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env bash
+
```
Comment (markus2330) [inline benchmarks/backend/heatmap/insert_data-small_map.sh:2]: Please describe what gets added. You can use this also in the test protocol.

**Review by markus2330 — CHANGES_REQUESTED** [review]
In general a lot of good things here but I could not review/verify many parts because documentation is missing. Please update documentation first so that it is clear which is for what and how to reproduce what you did in thesis/case study/etc.

File: benchmarks/backend/heatmap/setup.sh [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+# Insert data
+cd ../scraper && npm run insert
+printf "\nNow run the backend in a second shell.\n"
+read -n 1 -p "Press any key to continue!"
+../benchmarks/backend/heatmap/insert_data-large_map.sh $1 $2
+printf "\n\nStop the backend.\n"
+read -n 1 -p "Press any key to continue!"
+
+# Start the backend
+cd ../backend
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply benchmarks/backend/heatmap/setup.sh:19]: I don't like modifying files like this in scripts. If I were to append it to the Cargo.toml I would need to remove it again at the end of the script. However if I interrupt the script it might get messy, so I'd rather the user inserts it manually.

File: benchmarks/backend/heatmap/setup.sh [hunk 10]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply benchmarks/backend/heatmap/setup.sh:2]: Will be described in doc.

File: backend/Cargo.toml [hunk 11]
```diff
@@ -59,3 +59,6 @@ image = { version = "0.24.6", default-features = false, features = ["png"] }
 
 [dev-dependencies]
 jsonwebkey = { version = "0.3.5", features = ["generate", "jwt-convert"] }
+
+[profile.release]
+debug = true
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply backend/Cargo.toml:64]: I described it in doc. And yes this will be removed again (part of the reason why PR is draft), its just easier without having to exclude it from git every time.

File: benchmarks/backend/README.md [hunk 12]
```diff
@@ -0,0 +1,3 @@
+# Backend Benchmarks
+
+Documentation about the backend benchmarks can be found [here](../../doc/backend/06performance_benchmarks.md).
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply benchmarks/backend/README.md:3]: I put it into the doc so it is readable with mdbook. If put it here I could not link to it.

File: benchmarks/backend/config/setup.sh (hunk 7 above)
Comment (kitzbergerg) [reply benchmarks/backend/config/setup.sh:2]: General setup of backend and database. Described in doc.

File: benchmarks/backend/config/get_statistics.sh [hunk 13]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env nu
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply benchmarks/backend/config/get_statistics.sh:1]: For this one it technically isn't required, however I would like to keep it consistent.

File: benchmarks/backend/heatmap/get_statistics.sh [hunk 14]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env nu
+
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply benchmarks/backend/heatmap/get_statistics.sh:2]: Done in doc.

File: benchmarks/backend/config/run_httperf.sh [hunk 15]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply benchmarks/backend/config/run_httperf.sh:2]: Done in doc.

File: doc/tests/manual/protocol.md [hunk 16]
```diff
@@ -114,14 +114,15 @@
...
 - Preconditions:
-  - Be on the map managment page.
-  - TODO! @kitzbergerg
+  - Be on the map management page.
+  - Data is inserted via the scraper (plants and plant relations)
 - Test Steps:
   1. Create a map
   2. Plant some plants with relations.
-  3. TODO! @kitzbergerg
+  3. Add other constraints such as shade or soil ph.
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply doc/tests/manual/protocol.md:122]: Should this really be specific? I wrote it like this as we said we want to be general in this file. More specific details can be found in the reports.

**Comment by kitzbergerg** [comment]
jenkins build please

File: doc/tests/manual/reports/230723_heatmap_with_shade.md [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+
+```bash
+curl -o file.png --location 'http://localhost:8080/api/maps/1/layers/plants/heatmap?plant_id=1&plant_layer_id=1&shade_layer_id=3' \
+--header 'Authorization: Bearer <token>'
+```
+
+7. Verify:
+
+- The bottom left corner should be transparent, everything else should be green.
+- The top left corner should be green (as there is shade there and plant with id 1 likes shade); the rest should be yellow.
```
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply doc/tests/manual/reports/230723_heatmap_with_shade.md:238]: I don't understand what is unclear about this? This is a manual test, it doesn't have to represent actual data. The only thing that was tested there is if the shade is correctly considered for the plant.
As far as I understand the case study has nothing to do with manual tests.

File: benchmarks/backend/heatmap/insert_data-small_map.sh (hunk 8 above)
Comment (kitzbergerg) [reply benchmarks/backend/heatmap/insert_data-small_map.sh:2]: It is written in the script as comments. I don't want to use this for anything but the performance tests.
To use this for the test protocol would in my opinion be a misuse of the script. The manual test execution is already done, there is nothing that should now be changed about it (it points to a specific commit and works on that commit).

I will as a next step do the case study to verify if everything still works correctly and afterwards I will finish the benchmarks.

File: benchmarks/backend/heatmap/setup.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply benchmarks/backend/heatmap/setup.sh:19]: Yeah, without a proper TOML tool it is not a good idea. Elektra can do such things but I agree this is over-the-top for this :wink:

File: backend/Cargo.toml (hunk 11 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply backend/Cargo.toml:64]: Then add a TODO, then it is also clear for reviewers.

File: benchmarks/backend/config/get_statistics.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply benchmarks/backend/config/get_statistics.sh:1]: I would prefer shell/bash code, I didn't see a package for nu in Debian :cry:

File: doc/tests/manual/protocol.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply doc/tests/manual/protocol.md:121]: I agree in general but as long as we don't have enough experience how to test your code, it is better to be very specific to understand what you considered to be correct.

File: benchmarks/backend/config/get_statistics.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (kitzbergerg) [reply benchmarks/backend/config/get_statistics.sh:1]: You can see it here: https://www.nushell.sh/book/installation.html#installing-rust (I linked it in doc as well)
It is just way easier for me this way as I can easily execute groupings and parsing. It might be possible with bash as well, but that would take to long for me to figure out.

File: benchmarks/backend/heatmap/insert_data-small_map.sh (hunk 8 above)
Comment (markus2330) [reply benchmarks/backend/heatmap/insert_data-small_map.sh:2]: I don't see a problem in reusing the same scripts. Finally, we want also the manual execution to be automated.

> it points to a specific commit and works on that commit

That is fine as long as it also works for later commits.

> I will as a next step do the case study to verify if everything still works correctly and afterwards I will finish the benchmarks.

:+1:

File: benchmarks/backend/config/get_statistics.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply benchmarks/backend/config/get_statistics.sh:1]: Ahh, ok `cargo install nu` works, ok yes, then it is fine.

File: doc/tests/manual/reports/230723_heatmap_with_shade.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply doc/tests/manual/reports/230723_heatmap_with_shade.md:238]: For command 6 I get the error `Plant layer with id 1 not found`. It might be a problem on my side, a script would be easier.

But the main problem I have is that all the scaling in the head is not easy. It is not obvious which coordinates will be shown in the bottom left corner.

> As far as I understand the case study has nothing to do with manual tests.

(Manual) tests are constructed to be minimal and still test all interesting corner cases. A case study is realistic: it might not have all corner cases (would be a big plus, though, if it has) but has many different standard cases. Beside from that, they are more or less the same: they demonstrate that some code works.

**Comment by kitzbergerg** [comment]
@markus2330 @badnames This PR is now finished from my side including performance improvements, etc.

**Review by badnames — APPROVED** [review]

**Review by markus2330 — APPROVED** [review]
Fascinating work, especially with the extra effort that the frontend could not be used for putting shadings etc. A few pictures would have been very nice (for case study etc.) but okay in this case as it was done manually.

I'll keep the PR open until we can verify in the frontend that no changes in the backend are necessary.

File: backend/migrations/2023-07-03-165000_heatmap/up.sql [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+    score.preference := 1 / (1 + exp(-input.preference));       -- standard sigmoid, so that f(0)=0.5
+    score.relevance := (2 / (1 + exp(-input.relevance))) - 1;   -- modified sigmoid, so that f(0)=0
+    RETURN score;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Calculate score for a certain position.
+--
+-- p_map_id       ... map id
+-- p_layer_ids[1] ... plant layer
```
Comment (flo91) [inline backend/migrations/2023-07-03-165000_heatmap/up.sql:140]:
```suggestion
-- p_layer_ids[1] ... plant layer (only the first array-element is used by the function)
```

Alternatively, it could be more clear to change the array-argument to a value argument, e.g.
```SQL
p_layer_id ... id of the plant layer
```

File: backend/migrations/2023-07-03-165000_heatmap/up.sql [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+        weight := 1 / (1 + (distance / 50)^2);
+
+        -- update score based on relation
+        IF plant_relation.relation = 'companion' THEN
+            score.preference := score.preference + 0.5 * weight;
+            score.relevance := score.relevance + 0.5 * weight;
+        ELSE
+            score.preference := score.preference - 0.5 * weight;
+            score.relevance := score.relevance + 0.5 * weight;
+        END IF;
```
Comment (flo91) [inline backend/migrations/2023-07-03-165000_heatmap/up.sql:205]: As the 2nd calculation is the same in both branches, it could be moved outside of the if-else block to avoid code duplication and make it more obvious that the value of `relation` only affects `preference`, but not `relevance`.

```suggestion
        IF plant_relation.relation = 'companion' THEN
            score.preference := score.preference + 0.5 * weight;
        ELSE
            score.preference := score.preference - 0.5 * weight;
        END IF;
        score.relevance := score.relevance + 0.5 * weight;
```

File: backend/migrations/2023-07-22-110000_shadings/down.sql [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+    p_map_id INTEGER,
+    p_layer_ids INTEGER [],
+    p_plant_id INTEGER,
+    date DATE,
+    x_pos INTEGER,
+    y_pos INTEGER
+)
+RETURNS SCORE AS $$
+DECLARE
+    score SCORE;
```
Comment (flo91) [inline backend/migrations/2023-07-22-110000_shadings/down.sql:14]: The same as in [backend/migrations/2023-07-03-165000_heatmap/up.sql](https://github.com/ElektraInitiative/PermaplanT/pull/684/files#diff-90e920f3ee01f0da92f3e2f47880632407ab1bba770c44355e9bd0c864877896).

Is there any reason for using the variable `score` instead of directly returning `plants`?

**Review by flo91 — COMMENTED** [review]
Great and extensive work! 🎊 

Unfortunately, I'm not proficient in Rust and therefore I've focused my review on the SQL scripts.
I've not found any obvious bug, just some small suggestions and questions.

Thank you for this contribution! 👍🏻

**Comment by markus2330** [comment]
@flo91 thx for the review!

@badnames can you please work that review in (in your branch).

File: doc/backend/06performance_benchmarks.md [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+It will parse the PostgreSQL logs and httperf logs to extract execution times.
+
+## Example run
+
+The following is a step by step guide on how to execute the benchmark for the heatmap:
+
+1. Insert
+   ```toml
+    [profile.release]
+    debug = true
```
Comment (temmey) [inline doc/backend/06performance_benchmarks.md:88]: Doesn't a profile.bench make more sense here?

https://doc.rust-lang.org/cargo/reference/profiles.html#bench

Long-term, multiple build profiles could be defined in Cargo.toml.

**Review by temmey — CHANGES_REQUESTED** [review]

**Comment by badnames** [comment]
> @flo91 thx for the review!
> 
> @badnames can you please work that review in (in your branch).

Yes of course

File: backend/migrations/2023-07-22-110000_shadings/up.sql [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+DECLARE
+    score SCORE;
+    plants SCORE;
+    shades SCORE;
+BEGIN
+    plants := calculate_score_from_relations(p_layer_ids[1], p_plant_id, date, x_pos, y_pos);
+    shades := calculate_score_from_shadings(p_layer_ids[2], p_plant_id, date, x_pos, y_pos);
+
+    score.preference := plants.preference + shades.preference;
+    score.relevance := plants.relevance + shades.relevance;
```
[the remark this answers is not in the document]
Comment (badnames) [reply backend/migrations/2023-07-22-110000_shadings/up.sql:53]: Yes, but this shouldn't matter since the results will be clamped to 1 in the backend code.

**Comment by badnames** [comment]
@temmey @flo91 Your change requests have been incorporated into #804.

**Comment by markus2330** [comment]
I close now as there is already a follow-up PR.
